tests/components/mqtt/test_device_trigger.py

```diff
@@ -351,10 +371,191 @@ async def test_if_fires_on_mqtt_message(
+        '  "payload": "long_press",'
+        '  "topic": "foobar/triggers/button2",'
+        '  "type": "press",'
+        '  "subtype": "button" }'
+    )
+    async_fire_mqtt_message(hass, "homeassistant/device_automation/bla1/config", data1)
+    async_fire_mqtt_message(hass, "homeassistant/device_automation/bla2/config", data2)
+    await hass.async_block_till_done()
+    device_entry = device_registry.async_get_device(identifiers={("mqtt", "0AFFD2")})
+
+    assert await async_setup_component(
+        hass,
+        automation.DOMAIN,
+        {
+            automation.DOMAIN: [
+                {
+                    "trigger": {
+                        "platform": "device",
+                        "domain": DOMAIN,
+                        "device_id": device_entry.id,
+                        "type": "press",
+                        "subtype": "button",
+                    },
+                    "action": {
+                        "service": "test.automation",
+                        "data_template": {"some": ("press1")},
+                    },
+                },
+                {
+                    "trigger": {
+                        "platform": "device",
+                        "domain": DOMAIN,
+                        "device_id": device_entry.id,
+                        "type": "press",
+                        "subtype": "button",
+                    },
+                    "action": {
+                        "service": "test.automation",
+                        "data_template": {"some": ("press2")},
+                    },
+                },
+            ]
+        },
+    )
+
+    # Try to trigger first config.
+    # and triggers both attached instances.
+    async_fire_mqtt_message(hass, "foobar/triggers/button1", "short_press")
+    await hass.async_block_till_done()
+    assert len(calls) == 2
+    assert calls[0].data["some"] == "press1"
+    assert calls[1].data["some"] == "press2"
+
+    # Trigger second config references to same trigger
+    # and triggers both attached instances.
+    async_fire_mqtt_message(hass, "foobar/triggers/button2", "long_press")
+    await hass.async_block_till_done()
+    assert len(calls) == 2
+    assert calls[0].data["some"] == "press1"
+    assert calls[1].data["some"] == "press2"
+
+    # Removing the first trigger will clean up
+    calls.clear()
+    async_fire_mqtt_message(hass, "homeassistant/device_automation/bla1/config", "")
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+    assert (
+        "Device trigger ('device_automation', 'bla1') has been removed" in caplog.text
+    )
+    async_fire_mqtt_message(hass, "foobar/triggers/button1", "short_press")
+    assert len(calls) == 0
+
+
 async def test_if_fires_on_mqtt_message_template(
     hass: HomeAssistant,
     device_registry: dr.DeviceRegistry,
-    calls,
+    calls: list[ServiceCall],
     mqtt_mock_entry: MqttMockHAClientGenerator,
```

Comment:
To be consistent with other tests, typing for the `calls` fixture was added to all tests.